File: onion-architecture/java/src/test/java/com/example/api/EnrolmentControllerTest.java
package com.example.api;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import static org.mockito.ArgumentMatchers.any;
import org.mockito.Mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.test.web.servlet.MockMvc;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;

import com.example.application.EnrolmentApplicationService;

@ExtendWith(MockitoExtension.class)
class EnrolmentControllerTest {

    @Mock
    private EnrolmentApplicationService enrolmentApplicationService;

    private EnrolmentController controller;
    private MockMvc mockMvc;

    @BeforeEach
    void setUp() {
        controller = new EnrolmentController(enrolmentApplicationService);
        mockMvc = MockMvcBuilders.standaloneSetup(controller).build();
    }

    @Test
    void shouldEnrolStudentInCourseSuccessfully() throws Exception {
        Long studentId = 1L;
        Long courseId = 1L;

        mockMvc.perform(post("/enrolments")
                .param("studentId", studentId.toString())
                .param("courseId", courseId.toString()))
                .andExpect(status().isOk());

        verify(enrolmentApplicationService).enrol(studentId, courseId);
    }

    @Test
    void shouldAcceptValidStudentAndCourseIds() throws Exception {
        Long studentId = 123L;
        Long courseId = 456L;

        mockMvc.perform(post("/enrolments")
                .param("studentId", studentId.toString())
                .param("courseId", courseId.toString()))
                .andExpect(status().isOk());

        verify(enrolmentApplicationService).enrol(studentId, courseId);
    }

    @Test
    void shouldReturnBadRequestWhenStudentIdMissing() throws Exception {
        Long courseId = 1L;

        mockMvc.perform(post("/enrolments")
                .param("courseId", courseId.toString()))
                .andExpect(status().isBadRequest());

        verify(enrolmentApplicationService, never()).enrol(any(), any());
    }

    @Test
    void shouldReturnBadRequestWhenCourseIdMissing() throws Exception {
        Long studentId = 1L;

        mockMvc.perform(post("/enrolments")
                .param("studentId", studentId.toString()))
                .andExpect(status().isBadRequest());

        verify(enrolmentApplicationService, never()).enrol(any(), any());
    }

    @Test
    void shouldReturnBadRequestWhenBothParametersMissing() throws Exception {
        mockMvc.perform(post("/enrolments"))
                .andExpect(status().isBadRequest());

        verify(enrolmentApplicationService, never()).enrol(any(), any());
    }

    @Test
    void shouldReturnMethodNotAllowedForGetRequest() throws Exception {
        mockMvc.perform(get("/enrolments")
                .param("studentId", "1")
                .param("courseId", "1"))
                .andExpect(status().isMethodNotAllowed());

        verify(enrolmentApplicationService, never()).enrol(any(), any());
    }

    @Test
    void shouldCallServiceWithExactParameters() throws Exception {
        Long expectedStudentId = 42L;
        Long expectedCourseId = 99L;

        mockMvc.perform(post("/enrolments")
                .param("studentId", expectedStudentId.toString())
                .param("courseId", expectedCourseId.toString()))
                .andExpect(status().isOk());

        verify(enrolmentApplicationService).enrol(expectedStudentId, expectedCourseId);
        verifyNoMoreInteractions(enrolmentApplicationService);
    }
}

File: onion-architecture/kotlin/src/test/kotlin/com/example/api/EnrolmentControllerTest.kt
package com.example.api

import com.example.application.EnrolmentApplicationService
import io.mockk.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.BeforeEach
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.*
import org.springframework.test.web.servlet.setup.MockMvcBuilders

class EnrolmentControllerTest {

    private val enrolmentApplicationService = mockk<EnrolmentApplicationService>()
    
    private lateinit var controller: EnrolmentController
    private lateinit var mockMvc: MockMvc

    @BeforeEach
    fun setUp() {
        controller = EnrolmentController(enrolmentApplicationService)
        mockMvc = MockMvcBuilders.standaloneSetup(controller).build()
    }

    @Test
    fun shouldEnrolStudentInCourseSuccessfully() {
        val studentId = 1L
        val courseId = 1L
        
        every { enrolmentApplicationService.enrol(studentId, courseId) } just Runs
        
        mockMvc.perform(post("/enrolments")
                .param("studentId", studentId.toString())
                .param("courseId", courseId.toString()))
                .andExpect(status().isOk)
        
        verify { enrolmentApplicationService.enrol(studentId, courseId) }
    }

    @Test
    fun shouldAcceptValidStudentAndCourseIds() {
        val studentId = 123L
        val courseId = 456L
        
        every { enrolmentApplicationService.enrol(studentId, courseId) } just Runs
        
        mockMvc.perform(post("/enrolments")
                .param("studentId", studentId.toString())
                .param("courseId", courseId.toString()))
                .andExpect(status().isOk)
        
        verify { enrolmentApplicationService.enrol(studentId, courseId) }
    }

    @Test
    fun shouldReturnBadRequestWhenStudentIdMissing() {
        val courseId = 1L
        
        mockMvc.perform(post("/enrolments")
                .param("courseId", courseId.toString()))
                .andExpect(status().isBadRequest)
        
        verify(exactly = 0) { enrolmentApplicationService.enrol(any(), any()) }
    }

    @Test
    fun shouldReturnBadRequestWhenCourseIdMissing() {
        val studentId = 1L
        
        mockMvc.perform(post("/enrolments")
                .param("studentId", studentId.toString()))
                .andExpect(status().isBadRequest)
        
        verify(exactly = 0) { enrolmentApplicationService.enrol(any(), any()) }
    }

    @Test
    fun shouldReturnBadRequestWhenBothParametersMissing() {
        mockMvc.perform(post("/enrolments"))
                .andExpect(status().isBadRequest)
        
        verify(exactly = 0) { enrolmentApplicationService.enrol(any(), any()) }
    }

    @Test
    fun shouldReturnMethodNotAllowedForGetRequest() {
        mockMvc.perform(get("/enrolments")
                .param("studentId", "1")
                .param("courseId", "1"))
                .andExpect(status().isMethodNotAllowed)
        
        verify(exactly = 0) { enrolmentApplicationService.enrol(any(), any()) }
    }

    @Test
    fun shouldCallServiceWithExactParameters() {
        val expectedStudentId = 42L
        val expectedCourseId = 99L
        
        every { enrolmentApplicationService.enrol(expectedStudentId, expectedCourseId) } just Runs
        
        mockMvc.perform(post("/enrolments")
                .param("studentId", expectedStudentId.toString())
                .param("courseId", expectedCourseId.toString()))
                .andExpect(status().isOk)
        
        verify { enrolmentApplicationService.enrol(expectedStudentId, expectedCourseId) }
        verify(exactly = 1) { enrolmentApplicationService.enrol(any(), any()) }
    }
}